docs/config.md
# Configuration

You can use a JSON-encoded configuration file to overwrite the default settings
when running `antgen` optimizations. You only need to set changed values (while
honoring the nested structure) in a custom configuration. Do not include comments
in your configuration.

## "defaults"

Defaults for command-line options like leg length, wire specification, ground
parameters (see [NEC2 Ground card](https://nec2.org/part_3/cards/gn.html))
and source (feed) properties.

    {
        "defaults": {
            "k": 0.25,                     # leg length
            "wire": {                      # antenna wire
                "dia": 0.002,              #   diameter
                "G": 5.96e7,               #   conductivity (S/m)
                "L": 2e-8                  #   inductivity (H/m)
            },
            "ground": {                    # ground parameters
                "height": 0,               #   height over ground
                "mode": 0,                 #   ground type: 0=free-space
                "type": 0,
                "nradl": 0,
                "epse": 0,
                "sig": 0
            },
            "source": {                    # source properties
                "Z": {                     #   source impedance
                    "R": 50,
                    "X": 0
                },
                "power": 1,                # output power (W)
                "freq": 435000000,         # center frequency
                "span": 5000000            # frequency side range
            }
        },

## "simulation"

Simulation parameters control the flow of optimization, define terminating
conditions and the resolution of the radiation pattern.

        "simulation": {
            "maxRounds": 5,                 # terminate if not improvable
            "minZr": 1,                     # terminate if re(Z) too small
            "maxZr": 3000,                  # terminate if re(Z) too big
            "minChange": 0.001,             # terminate if progress is too small
            "progressCheck": 10,            # check progress every 10 iterations
            "minBend": 0.01,                # min. bend is 1% of max. bend
            "exciteU": 1.0,                 # excitation voltage
            "phiStep": 5.0,                 # resolution of RP in elevation
            "thetaStep": 5.0,               # resolution of RP in azimuth
            "wireMax": 0.008,               # max. wire diameter in λ
            "segMinLambda": 0.002,          # min. segment length in λ
            "segMinWire": 4,                # segment at least 4 wire diameters
            "minRadius": 0.02               # smallest bend radius (in λ)
        },

## "material"

Pre-defined wire material parameters:

        "material": {
            "Cu": {
                "conductivity": 5.96e7,
                "inductance": 1.32e-6
            },
            "CuL": {
                "conductivity": 5.96e7,
                "inductance": 1.1e-7
            },
            "Al": {
                "conductivity": 3.5e7,
                "inductance": 2.5e-8
            }
        },

## "plugins"

As no evaluator plugins are built-in, this section is usually empty. If you
have build your own plugin, you can add it to your configuration file:

        "plugins": {
            "mytarget": "./mytarget_evaluator.so"
        },

The advantage of having the plugin in the configuration file is that using
the plugin in `antgen` is simplified, because you can reference the plugin
by name:

     antgen ... -opt plugin:@mytarget

## "render"

This section defines the properties of the render engine/window:

        "render": {
            "canvas": "sdl",                 # use SDL for rendering
            "width": 1024,                   # width of render window
            "height": 768                    # height of render window
        }
    }


lib/config.go
//----------------------------------------------------------------------
// This file is part of antgen.
// Copyright (C) 2024-present Bernd Fix >Y<,  DO3YQ
//
// antgen is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License,
// or (at your option) any later version.
//
// antgen is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// SPDX-License-Identifier: AGPL3.0-or-later
//----------------------------------------------------------------------

package lib

import (
	"encoding/json"
	"os"
)

// Simulation parameters
type Simulation struct {
	// optimization parameters (termination conditions)
	MaxRounds     int     `json:"maxRounds"`     // max. number of rounds in optimization
	MinZr         float64 `json:"minZr"`         // min. resistance of antenna
	MaxZr         float64 `json:"maxZr"`         // max. resistance of antenna
	MinChange     float64 `json:"minChange"`     // progress check: min. change in target value
	ProgressCheck int     `json:"progressCheck"` // number of steps between progress check
	MinBend       float64 `json:"minBend"`       // min. bending angle (fraction of max. angle)

	// simulation-related constants (NEC2 simulation)
	ExciteU   float64 `json:"exciteU"`   // excitation voltage
	PhiStep   float64 `json:"phiStep"`   // azimut step (degree)
	ThetaStep float64 `json:"thetaStep"` // elevation step (degree)

	// geometry-related constraints (NEC2 simulation)
	WireMax      float64 `json:"wireMax"`      // max. wire diameter (in wavelength)
	SegMinLambda float64 `json:"segMinLambda"` // min. segment length (in wavelength)
	SegMinWire   float64 `json:"segMinWire"`   // min. segment length (in wire diameter)
	MinRadius    float64 `json:"minRadius"`    // min. curve radius (in wavelength)
}

// Material spec for wires
type Material struct {
	Conductivity float64 `json:"conductivity"` // wire conductivity (S/m)
	Inductance   float64 `json:"inductance"`   // wire inductance (H/m)
}

// RenderConfig for rendering-related settings
type RenderConfig struct {
	Canvas string `json:"canvas"` // render engine/canvas
	Width  int    `json:"width"`  // width of canvas (usually in pixels)
	Height int    `json:"height"` // height of canvas (usually in pixels)
}

// Config for AntGen
type Config struct {
	Def     *Specification       `json:"default"`
	Sim     *Simulation          `json:"simulation"`
	Mat     map[string]*Material `json:"material"`
	Render  *RenderConfig        `json:"render"`
	Plugins map[string]string    `json:"plugins"`
}

// Cfg is the globally-accessible configuration (pre-set)
var Cfg = &Config{
	// default values (command-line options)
	Def: &Specification{
		K: 0.25,
		Wire: Wire{
			Diameter:     0.002,
			Material:     "",
			Conductivity: 0,
			Inductance:   0,
		},
		Ground: Ground{
			Height: 0,
			Mode:   0,
			Type:   -1,
			NRadl:  0,
			Epse:   0,
			Sig:    0,
		},
		Source: Source{
			Z:     Impedance{50, 0},
			Power: 1,
			Freq:  435000000,
			Span:  5000000,
		},
		Feedpt: Feedpt{
			Gap:       0,
			Extension: 0,
		},
	},
	// Simulation parameters
	Sim: &Simulation{
		// optimization parameters (termination conditions)
		MaxRounds:     5,
		MinZr:         1,
		MaxZr:         20000,
		MinChange:     0.001,
		ProgressCheck: 10,
		MinBend:       0.01,

		// simulation-related constants (NEC2 simulation)
		ExciteU:   1.0,
		PhiStep:   5.0,
		ThetaStep: 5.0,

		// geometry-related constraints (NEC2 simulation)
		WireMax:      0.008,
		SegMinLambda: 0.002,
		SegMinWire:   4,
		MinRadius:    0.02,
	},
	// rendering parameters
	Render: &RenderConfig{
		Canvas: "sdl",
		Width:  1024,
		Height: 768,
	},
	// wire materials
	Mat: map[string]*Material{
		"Cu": { // cupper wire
			5.96e7,  // conductivity (S/m)
			1.32e-6, // inductance (H/m)
		},
		"CuL": { // enamel copper wire
			5.96e7, // conductivity (S/m)
			1.1e-7, // inductance (H/m)
		},
		"Al": { // Aluminium
			3.5e7,  // conductivity (S/m)
			2.5e-8, // inductance (H/m)
		},
	},
	// no pre-defined plugins
	Plugins: make(map[string]string),
}

// ReadConfig from file
func ReadConfig(fname string) (err error) {
	var data []byte
	if data, err = os.ReadFile(fname); err == nil {
		err = json.Unmarshal(data, &Cfg)
	}
	return
}
